tests/Integration/Behaviour/Features/Context/Domain/Product/AbstractProductFeatureContext.php

```diff
@@ -116,6 +117,48 @@ protected function assertStringProperty(ProductForEditing $productForEditing, ar
         }
     }
 
+    /**
+     * @param ProductForEditing $productForEditing
+     * @param array $data
+     * @param string $propertyName
+     */
+    protected function assertNumberProperty(ProductForEditing $productForEditing, array &$data, string $propertyName): void
+    {
+        if (isset($data[$propertyName])) {
+            $expectedValue = new DecimalNumber((string) $data[$propertyName]);
+            $actualValue = new DecimalNumber((string) $this->extractValueFromProductForEditing($productForEditing, $propertyName));
+
+            Assert::assertTrue(
+                $expectedValue->equals($actualValue),
+                sprintf('Expected %s %s. Got %s.', $propertyName, (string) $expectedValue, (string) $actualValue)
+            );
+
+            unset($data[$propertyName]);
+        }
+    }
+
+    /**
+     * @param ProductForEditing $productForEditing
+     * @param array $data
+     * @param string $propertyName
+     */
+    protected function assertDateProperty(ProductForEditing $productForEditing, array &$data, string $propertyName): void
+    {
+        if (isset($data[$propertyName])) {
+            $expectedValue = PrimitiveUtils::castElementInType($data[$propertyName], PrimitiveUtils::TYPE_DATETIME);
+            $actualValue = $this->extractValueFromProductForEditing($productForEditing, $propertyName);
+            $formattedExpectedDate = $expectedValue->format('Y-m-d');
+            $formattedActualDate = $actualValue->format('Y-m-d');
+            Assert::assertEquals(
+                $formattedExpectedDate,
+                $formattedActualDate,
+                sprintf('Expected %s "%s". Got "%s".', $propertyName, $formattedExpectedDate, $formattedActualDate)
+            );
+
+            unset($data[$propertyName]);
```

Comment:
I understand this is done because of
```
        // Assertions checking isset() can hide some errors if it doesn't find array key,
        // to make sure all provided fields were checked we need to unset every asserted field
        // and finally, if provided data is not empty, it means there are some unnasserted values left
```

The 1st time I checked this function I was like "wtf ? 🤔 " when I saw the unset. Now I understand.

Could we make it easier for the reader to understand this ? My suggestions:
- store this statement/explanation somewhere we can lead people to (phpDoc of this class ? devdocs ?) and refer to it
- add into `assertDateProperty()` phpDoc the explanation about the isset (can be a simple link)
- modify `assertDateProperty()` into something more meaningful `assertDatePropertyAndUnsetData()` but it's very verbose, I dont like it

Reply:
All the assert methods from this class use the same principle of unsetting checked value I was surprised first time I saw that myself so I understand your trouble

Though since they all work the same I wouldn't go for the `assertDatePropertyAndUnsetData` which can become unreadable if they are all renamed like this

I'd rather go for a simple comment on top of each `unset`